### ISSUE TITLE
Investigate performance optimization of model accessors

### ISSUE THREAD

**Issue by notbakaneko** [ppy/osu-web#4213]
The way Eloquent model attribute values are accessed and casted seem to incur a significant overhead. Considering how fundamental and frequent getting and setting model attributes are, it would be really good not to be constantly be hit with this overhead.

This is part of a related series looking at the performance of osu-web and the same endpoints are used:
#4207
#4212

In this part we are going to look at making optimizations at how attribute values are casted, comparing the following cases:

1. Default accessors (base)
2. Default accessors with attribute caching
3. Optimized castAttribute() usage with attribute caching
4. Direct getAttribute without castAttribute() with attribute caching
5. Direct getAttribute without castAttribute() without attribute caching

Attribute caching is the same caching as being tested in https://github.com/ppy/osu-web/issues/4207

For the test, we modified `Beatmap` and `Beatmapset` to make use of the optimizations.
Summary of the results (requests per second):

  | Page size | Case 1 | Case 2 | Case 3 | Case 4 | Case 5
-- | -- | -- | -- | -- | -- | --
/home |   | 807.39 | 802.48 | 802.44 | 804.43 | 810.21
/beatmapsets/search | 1 | 600.88 | 601.37 | 601.48 | 602.31 | 607.75
  | 10 | 446.58 | 456.54 | 466.56 | 475.23 | 475.12
  | 25 | 306.16 | 320.15 | 330.02 | 346.40 | 343.44
  | 50 | 198.91 | 209.95 | 222.28 | 238.12 | 235.14
  | 100 | 114.29 | 121.50 | 130.21 | 141.84 | 139.37
  | 150 | 81.87 | 83.56 | 89.98 | 98.21 | 100.89

Or performance relative to base (1):

  | Page size | Case 1 | Case 2 | Case 3 | Case 4 | Case 5
-- | -- | -- | -- | -- | -- | --
/home |   | 100.00% | 99.39% | 99.39% | 99.63% | 100.35%
/beatmapsets/search | 1 | 100.00% | 100.08% | 100.10% | 100.24% | 101.14%
  | 10 | 100.00% | 102.23% | 104.47% | 106.42% | 106.39%
  | 25 | 100.00% | 104.57% | 107.80% | 113.15% | 112.18%
  | 50 | 100.00% | 105.55% | 111.75% | 119.71% | 118.22%
  | 100 | 100.00% | 106.31% | 113.93% | 124.11% | 121.95%
  | 150 | 100.00% | 102.06% | 109.91% | 119.96% | 123.24%

There is another case that uses the more familiar parameterless `getProperty()` style of getter common in other OO-languages, but the results are not included here; for reference, the difference is within 1% of case 4.

### Default accessors
This is the default, using Laravel’s attribute lookup and casting.

### Default accessors with attribute caching
Same case as https://github.com/ppy/osu-web/issues/4207

### Optimized castAttribute() usage with attribute caching
For this case, `Model::getAttributeValue()` is overridden and the lookups and conversions that are not relevant to our usage are removed.

In this scenario, there is about a 11% increase in performance over the base case with the standard beatmapset search page size and about 5% improvement over just having caching; there are potentially substantial performance benefits for not using Laravel’s in-built casting.

### Direct getAttribute without castAttribute() with attribute caching
For this case, all attribute casts lookups are removed; Laravel’s casting is bypassed; the flow when an attribute is accessed is now:
1. `model->key`
2. `model->__get($key)`
3. `model->getAttribute($key)`
4.`model->getAttributeValue($key)`
5. `model->getKeyAttribute(...)`

There are no longer any cast or type lookup functions; all casting is performed directly by the attribute mutators. In addition, this is combined with attribute caching so the attribute values do not have to be casted repeatedly on access.

In scenarios where model attributes are frequently accessed, there is potentially some sizeable performance benefits:

  | Page size | Case 4
-- | -- | --
/home |   | 99.63%
/beatmapsets/search | 1 | 100.24%
  | 10 | 106.42%
  | 25 | 113.15%
  | 50 | 119.71%
  | 100 | 124.11%
  | 150 | 119.96%

### Direct getAttribute without castAttribute() without attribute caching
This is the same as above case, but without attribute caching, so accessing the attribute again will require the value to be casted again.

  | Page size | Case 5
-- | -- | --
/home |   | 100.35%
/beatmapsets/search | 1 | 101.14%
  | 10 | 106.39%
  | 25 | 112.18%
  | 50 | 118.22%
  | 100 | 121.95%
  | 150 | 123.24%

Even without attribute caching, this scenario is still substantially more performant than:
- The base case
- Using attribute caching only
- Optimized castAttribute() usage with attribute caching

The main stumbling block here would be generating the necessary attribute getters and setters since it would be very inconvenient to do manually. Ideally, we would want to be able to specify the casts declaratively and either:
- have the appropriate methods pre-generated and bootstrapped somehow.
- have the appropriate methods generated at runtime via reflection; the issue with this is we don't want to be running the generation code on every request.

tl;dr; we could be 20% more performant without Laravel's attribute casting.

**Comment by nanaya** [comment]
Attribute casting has been overridden for some highly used models through `getAttribute` function.
